Improve Javadoc creation

Marcono1234 opened the pull request:
Resolves #522

Proposed commit message:

```
Improve Javadoc creation (#522 / #566)

Resolves: #522
PR: #566
```

Improves Javadoc creation by:
- Using [Gradle toolchains](https://docs.gradle.org/current/userguide/toolchains.html) to create the Javadoc with JDK 17
- Add external links to the JUnit Javadoc
- (Implicitly) add external links to the JDK Javadoc
- Enable doclint, except for the `missing` group

Future suggested work:
I currently do not completely understand why the project is built multiple times with different target Java versions and with different modularity settings. In my opinion (but maybe I am overlooking something):
- The project should always be built modular, but 

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [x] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code
* [x] Code adheres to code style, naming conventions etc.
* [x] Successful tests cover all changes
* [x] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [x] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [x] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

Marcono1234 commented on build.gradle.kts:
```diff
@@ -271,3 +287,10 @@ tasks {
 	}
 }
+
+tasks.withType<Javadoc>().configureEach {
```
Ideally this would be part of the `javadoc` task definition above, but then it failed to find functions, possibly because the 'wrong' `this` was in scope.
Would be great if someone more familiar with the Gradle Kotlin DSL could give it a try.

Alternatively the complete `javadoc` block above could be put inside `tasks.javadoc { ... }` including the toolchain definition below, but then the tasks configurations are separated which might decrease readability.

Marcono1234 replied:
Thanks, that worked!

Michael1993 commented on build.gradle.kts:
```diff
@@ -225,12 +227,26 @@ tasks {
+			links = listOf("https://junit.org/junit5/docs/current/api/")
+
+			// Set javadoc `--release` flag (affects which warnings and errors are reported)
+			addStringOption("-release", targetJavaVersion.getMajorVersion())
```
Comment doesn't seem to match code here.

Marcono1234 replied:
Do you mean because of `--` vs. `-`? Gradle seems to prepend the options with a `-`, so `-release` becomes `--release`.
`javadoc` [expects `--release`](https://docs.oracle.com/en/java/javase/17/docs/specs/man/javadoc.html#options-for-javadoc) which is why I wrote that in the comment.

Marcono1234 replied:
Also feel free to test the changes locally. For example without the `--release` option, if one of the classes would for example contain `private int _;`, a `javadoc` error would be shown because `_` is a reserved identifier for Java 17.
However, with `--release 8` option, it only causes a warning because Java 8 still permits it.

Michael1993 commented on build.gradle.kts:
```diff
@@ -225,12 +227,26 @@ tasks {
+			// https://docs.oracle.com/en/java/javase/17/docs/specs/man/javadoc.html#additional-options-provided-by-the-standard-doclet
+			// The Gradle option methods are rather misleading, but a boolean `true` value just makes sure the flag
+			// is passed to javadoc, see https://github.com/gradle/gradle/issues/2354
+			addBooleanOption("Xdoclint:all,-missing", true)
```
I wonder if the same is true here... Is this `--missing` or just `-missing`?

Marcono1234 replied:
It is `-missing`. That is part of the `-Xdoclint` option, it is not a standalone option.
See https://docs.oracle.com/en/java/javase/17/docs/specs/man/javadoc.html#additional-options-provided-by-the-standard-doclet (also linked in the comment in the source above)

Marcono1234 replied:
Also feel free to test the changes locally. Only using `Xdoclint:all` will show the undesired warnings about missing javadoc tags (e.g. for parameters). With `-missing` it only shows the other warnings, currently that the code does not use modules but links against the JUnit doc which uses modules, that can be ignored and could also be solved in the future with #567.

beatngu13 commented on build.gradle.kts:
```diff
@@ -225,12 +227,27 @@ tasks {
+			// https://docs.oracle.com/en/java/javase/17/docs/specs/man/javadoc.html#additional-options-provided-by-the-standard-doclet
+			// The Gradle option methods are rather misleading, but a boolean `true` value just makes sure the flag
+			// is passed to javadoc, see https://github.com/gradle/gradle/issues/2354
+			addBooleanOption("Xdoclint:all,-missing", true)
```
Why `addBooleanOption` instead of `addStringOption`?

Also, the [Javadoc manual](https://docs.oracle.com/en/java/javase/17/docs/specs/man/javadoc.html#additional-options-provided-by-the-standard-doclet) says:

```text
-Xdoclint:(all|none|[-]group)
```

I'm not that familiar with the Javadoc CLI, but does it really make sense to combine `all` and `missing` (especially with a leading `-`)?

```diff
-Xdoclint:accessibility,html,syntax,reference
+Xdoclint:all,-missing
```

Marcono1234 replied:
> Why `addBooleanOption` instead of `addStringOption`?

This is explained a bit in the Gradle issue linked in the comment; the name and behavior of these methods is rather confusing. Apparently Gradle considers a "boolean option" an option without separate value, e.g. `-option`. Whereas a "string option" would be an option with separate value, e.g. `-option value`. Since the doclint setting is part of the same argument it has to be a "boolean option".
However, using a "string option" with `null` value (explicitly or when calling the `addStringOption` method without separate value parameter), the option seems to be simply omitted from the command line arguments (see [Gradle implementation](https://github.com/gradle/gradle/blob/fc26c4c3e3edb3ef3572a4b663fc452689f2ad19/subprojects/language-java/src/main/java/org/gradle/external/javadoc/internal/StringJavadocOptionFileOption.java#L28-L30)).

---

> I'm not that familiar with the Javadoc CLI, but does it really make sense to combine `all` and `missing` (especially with a leading `-`)?

The leading `-` excludes the group. I tried to explain that with the comment "Enable doclint, but ignore warnings for missing tags" a few lines above, but maybe that was not obvious; do you have any suggestions for improving the comment?
This configuration is mentioned by the following bullet point in the linked documentation:
> <code>-Xdoclint all,-<i>group</i></code>: Enables all checks except <i>group</i> checks

beatngu13 replied:
Thank you for that great explanation!

beatngu13 commented on build.gradle.kts:
```diff
@@ -225,12 +227,27 @@ tasks {
+			// The Gradle option methods are rather misleading, but a boolean `true` value just makes sure the flag
+			// is passed to javadoc, see https://github.com/gradle/gradle/issues/2354
+			addBooleanOption("Xdoclint:all,-missing", true)
+			encoding = "UTF-8"
```
I would move this up just before `links`. Then, the `options` block becomes:

```kotlin
this as StandardJavadocDocletOptions

encoding = "UTF-8"
links = listOf("https://junit.org/junit5/docs/current/api/")

// options
// ...
```

IMHO a bit easier to read, what do you think?

beatngu13 reviewed (changes requested):
Another great PR, thank you!

beatngu13 commented:
@Marcono1234 after adding the `merge-ready` label, which enables the `full-featured` matrix build, Javadoc generation fails with the modular build on Java 17:

https://github.com/junit-pioneer/junit-pioneer/runs/4731184102?check_suite_focus=true

Would you mind having a look?

Marcono1234 commented:
It looks like the issue is:
>  \> Task :javadoc FAILED
> error: option --module-path not allowed with target 8

The issue occurred when a modular build (`-PmodularBuild=true`) was performed. I have changed the Javadoc task to now always at least use Java 9 as `--release` version. However, that has the disadvantage that exclusion of packages does not work anymore (I have commented it out). The proper solution would likely be to use the `-exclude` Javadoc option; not to be confused with Gradle's `exclude` functions which are currently used in the script and which exclude source files completely (causing Javadoc failures for modular builds due to unresolved imports), instead of just skipping Javadoc generation for them. Unfortunately using `-exclude` does not seem to work due to how Gradle invokes the Javadoc tool, see comments I added to the build script. `-exclude` would probably only work if Gradle used `-subpackages`, but it does not use that at the moment.

I have also updated JaCoCo because the previous version did not support Java 17 which caused errors to be printed to the console. Though if you want I can also create a separate pull request for that.

Michael1993 commented on build.gradle.kts:
```diff
@@ -93,11 +95,11 @@ yamlValidator {
 
 sonarqube {
-	// If you want to use this logcally a sonarLogin has to be provide, either via Username and Password
+	// If you want to use this logically a sonarLogin has to be provided, either via Username and Password
```
Wait, isn't this supposed to be "locally"? 😅

Marcono1234 replied:
Ah, that makes a lot more sense 😅
Have fixed it now, sorry.

beatngu13 commented on build.gradle.kts:
```diff
@@ -225,12 +227,37 @@ tasks {
+		/*
+		 * TODO: Currently disabled because modular Gradle build fails when this package is excluded
+		 * See https://stackoverflow.com/q/32785002 and slightly related Gradle issue https://github.com/gradle/gradle/issues/14066
+		*/
```
Missing alignment

```suggestion
		 */
```

Marcono1234 replied:
Thanks, have fixed it.

beatngu13 commented on build.gradle.kts:
```diff
@@ -225,12 +227,37 @@ tasks {
+		 * TODO: Currently disabled because modular Gradle build fails when this package is excluded
+		 * See https://stackoverflow.com/q/32785002 and slightly related Gradle issue https://github.com/gradle/gradle/issues/14066
+		*/
+		//exclude("org/junitpioneer/internal")
```
Would this help?

```suggestion
		if (!modularBuild.toBoolean()) {
			exclude("org/junitpioneer/internal")
		}
```

Marcono1234 replied:
Yes that works, and is probably a better solution for now, have changed this.

beatngu13 commented:
Thanks for your perseverance – great PR!
